core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -128,53 +133,105 @@ public class JsPeripheral internal constructor(
     private var connectJob: Deferred<Unit>? = null
 
     private fun connectAsync() = scope.async(start = LAZY) {
-        logger.info { message = "Connecting" }
-        _state.value = State.Connecting.Bluetooth
-
         try {
-            registerDisconnectedListener()
-
-            gatt.connect().await() // todo: Catch appropriate exception to emit State.Rejected.
-            _state.value = State.Connecting.Services
-
-            discoverServices()
-            onServicesDiscovered(ServicesDiscoveredPeripheral(this@JsPeripheral))
-            _state.value = State.Connecting.Observes
-            logger.verbose { message = "Configuring characteristic observations" }
-            observers.onConnected()
+            openConnection()
         } catch (t: Throwable) {
             logger.error(t) { message = "Failed to connect" }
-            disconnectGatt()
+            scope.async {
+                disconnectSync(cancelConnectJob = false)
```

Comment:
I don't follow why it is desirable to _not_ cancel the connect job when library consumer requested a `disconnect`?

Reply:
This is inside the connect job itself, and this catch is the path where it is already terminating so there is no need to cancel itself. It also avoid recursively canceling itself which makes for a somewhat confusing stack trace.